server.js

```diff
@@ -1,39 +1,220 @@
-// import booksData from './data/books.json'
-// import netflixData from './data/netflix-titles.json'
-// import topMusicData from './data/top-music.json'
+import topMusicData from './data/top-music.json'
+
+dotenv.config()
 
 const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/project-mongo"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
+mongoose.connect(mongoUrl, {
+  useNewUrlParser: true,
+  useUnifiedTopology: true
+})
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
+const musicTrackSchema = new mongoose.Schema({
+  id: Number,
+  trackName: {
+    type: String,
+    lowercase: true
+  },
+  artistName: {
+    type: String,
+    lowercase: true
+  },
+  genre: {
+    type: String,
+    lowercase: true
+  },
+  bpm: Number,
+  energy: Number,
+  danceability: Number,
+  loudness: Number,
+  liveness: Number,
+  valence: Number,
+  length: Number,
+  acousticness: Number,
+  speechiness: Number,
+  popularity: Number
+})
+
+const MusicTrack = mongoose.model('MusicTrack', musicTrackSchema)
+
+if (process.env.RESET_DB) {
+  const seedDB = async () => {
+    await MusicTrack.deleteMany()
+
+    await topMusicData.forEach(item => {
+      const NewMusicTrack = new MusicTrack(item)
+      NewMusicTrack.save()
+    })
+  }
+  seedDB()
+}
+
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
 app.use(cors())
-app.use(bodyParser.json())
+app.use(express.json())
+
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).json({
+      error: `Service unavailable`
+    })
+  }
+})
 
-// Start defining your routes here
+//root
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send('Popular Music API')
+})
+
+//all tracks, all info + query trackname and artist
+app.get('/tracks', async (req, res) => {
+  const { trackname, artist } = req.query
+
+  try {
+    const tracks = await MusicTrack.aggregate([{
+      $match: {
+        trackName: {
+          $regex: new RegExp(trackname || '', 'i')
+        },
+        artistName: {
+          $regex: new RegExp(artist || '', 'i')
+        }
+      }
+    }])
+
+    if (tracks.length) {
+      res.json({ length: tracks.length, data: tracks })
+    } else {
+      res.status(404).send({ message: `No match found` })
+    }
+  } catch {
+    res.status(400).json({ error: `Invalid request` })
+  }
+})
+
+//sorts all tracks based on popularity
+app.get('/tracks/popularity', async (req, res) => {
+
+  try {
+    let tracks = await MusicTrack.find()
+    tracks = tracks.sort((a, b) => b.popularity - a.popularity)
+    res.json({ length: tracks.length, data: tracks })
+  } catch {
+    res.status(400).json({ error: `Invalid request` })
+  }
+
```

Comment:
Nice filtering! 👌 If it's a very large database it could be useful to not show all of them, maybe just a top 10 or 25? Not a problem in this project though. ;-) An extra thing could be to show 20 per page or something.